include/circt/Dialect/HW/HWTypes.td

```diff
@@ -170,6 +170,20 @@ def TypeAliasType : HWType<"TypeAlias"> {
   }];
 }
 
+// Type constraint that supports a type alias or an inner type.
+class TypeAliasOrConstraint<Type innerType> : TypeConstraint<Or<[
```

Comment:
Parsing this is making me dizzy... Can you add newlines and indents to help with the nesting?